core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -316,7 +332,12 @@ public void addSplit(RemoteSplit split)
                 ExchangeDataSource dataSource = delegate.get();
                 if (dataSource == null) {
                     if (split.getExchangeInput() instanceof DirectExchangeInput) {
-                        DirectExchangeClient client = directExchangeClientSupplier.get(systemMemoryContext, taskFailureListener, retryPolicy);
+                        DirectExchangeClient client = directExchangeClientSupplier.get(
+                                taskId.getQueryId(),
+                                new ExchangeId(format("direct-exchange-%s-%s", taskId.getStageId().getId(), sourceId)),
```

Comment:
Why do we need to include `sourceId` here? If I remember correctly, there will be only one deduplication buffer on the coordinator?

Reply:
Because a single stage may be consuming more than a single source

Reply:
Does it hold for stages not involving coordinators? I don't see you include `sourceId` there.

Reply:
This is somehow confusing to be honest. The external exchanges are named after the producing stage, that is always unique. Direct exchanges named after a consuming stage (that may consume data from multiple producing stages). Naming a direct exchange after a producing stage is not possible, as a single consumer may consume data from multiple source stages.